pyanaconda/payload/dnfpayload.py

```diff
@@ -412,80 +414,149 @@ def addRepo(self, ksrepo):
+
+        kernel_package = self._get_kernel_package()
+        if kernel_package:
+            selected_packages.add(kernel_package)
+
+        # resolve packages and groups required by Anaconda
+        self.requirements.apply()
 
-        self._select_kernel_package()
+        # let's piece it all together and feed it to DNF at once
+        include_set = set()
+        exclude_set = set()
+        # note about formatting:
+        # - leading @ signifies a group or module
+        # - no leading @ means a package
+
+        # add environment
+        if env:
+            include_set.add("@^{}".format(env))
+
+        # add groups
+        include_set.update(group_specs)
+        for group_name in self._req_groups:
+            include_set.add("@{}".format(group_name))
+        # add packages
+        include_set.update(selected_packages)
+        include_set.update(self._req_packages)
+
+        # exclude groups
+        exclude_set.update(excluded_group_specs)
+        # exclude packages
+        exclude_set.update(excluded_packages)
+
+        # log the resulting set
+        log.debug("transaction include set")
+        log.debug(include_set)
+        log.debug("transaction exclude set")
+        log.debug(exclude_set)
+
+        # feed it to DNF
+        try:
+            self._base.install_specs(install=include_set, exclude=exclude_set)
+        except Exception as e:  # pylint: disable=broad-except
+            exception_message = str(e)
+            self._payload_setup_error(e)
 
     def _apply_requirements(self, requirements):
+        self._req_groups = set()
+        self._req_packages = set()
         for req in self.requirements.packages:
             ignore_msgs = []
             if req.id in self.instclass.ignoredPackages:
                 ignore_msgs.append("IGNORED by install class %s" % self.instclass)
             if req.id in self.data.packages.excludedList:
                 ignore_msgs.append("IGNORED because excluded")
             if not ignore_msgs:
-                try:
-                    self._install_package(req.id, required=req.strong)
-                except payload.NoSuchPackage as e:
-                    self._miss(e)
+                # NOTE: req.strong not handled yet
+                self._req_packages.add(req.id)
             log.debug("selected package: %s, requirement for %s %s",
                        req.id, req.reasons, ", ".join(ignore_msgs))
 
         for req in self.requirements.groups:
-            try:
-                self._select_group(req.id, required=req.strong)
-                log.debug("selected group: %s, requirement for %s",
-                           req.id, req.reasons)
-            except payload.NoSuchGroup as e:
-                self._miss(e)
+            # NOTE: req.strong not handled yet
+            log.debug("selected group: %s, requirement for %s",
+                       req.id, req.reasons)
+            self._req_groups.add(req.id)
```

Comment:
Please, add here the group specification, so you don't have to do it later.

Reply:
As mentioned, I'd prefer to do the requirements handling adjustment in a followup PR.